### DISCUSSION TITLE
How can I use per_species_rescale_shifts with exact values?

### DISCUSSION THREAD

**Discussion by pobo95** [mir-group/nequip#406] [Q&A] [answered]
Thank you for your good code!

These days, I have a problem of becoming gas on the melting point in NPT ensmeble.

And I'm thinking about `per_species_rescale_shifts` to solve this problem.

`dataset_per_atom_total_energy_mean` the default option does not seem to be a physical value.
Thats why I want to use exact values.

How can I use `per_species_rescale_shifts` with exact values?

When I use

> global_rescale_shift: null
global_rescale_shift_trainable: false
global_rescale_scale: dataset_forces_rms
global_rescale_scale_trainable: false
per_species_rescale_shifts_trainable: false
per_species_rescale_scales_trainable: true
per_species_rescale_shifts: [Ge:-0.02859140,Sb:-0.03305661,Te:-0.03357790]
per_species_rescale_scales: dataset_per_species_forces_rms

It showed this `KeyError: 'per_species_rescale_arguments_in_dataset_units'` massage.

And when I used `dataset_per_species_total_energy_mean` option, it doen't work with lammps
Do you have any suggestions?

Thank you for your help.

**Comment by Linux-cpp-lisp** [comment, marked as answer]
Hi @pobo95 ,

Thanks for your interest in our code!

You need to set `per_species_rescale_arguments_in_dataset_units: true` to tell the code that the values you are providing for the `per_species_rescale_shifts` are in the same units of the dataset, rather than the scaled units of the model.  Does it work if you set this value?

**Reply by pobo95** [reply]
Thank you for your quick answer @Linux-cpp-lisp!

No it doesn't work.

When I try 

> global_rescale_shift: null
global_rescale_shift_trainable: false
global_rescale_scale: dataset_forces_rms
global_rescale_scale_trainable: false
per_species_rescale_shifts_trainable: false
per_species_rescale_arguments_in_dataset_units: true
per_species_rescale_scales_trainable: true
per_species_rescale_shifts: [Ge: -0.02859140,Sb: -0.03305661,Te: -0.03357790]
per_species_rescale_scales: dataset_per_species_forces_rms


It got

> Successfully loaded the data set of type ASEDataset(2000)...
Replace string dataset_per_species_forces_rms to tensor([0.4694, 0.6948, 0.6315], dtype=torch.float64)
Traceback (most recent call last):
  File "/home/grad/pjpark/anaconda3/envs/nequip/bin/nequip-train", line 8, in <module>
    sys.exit(main())
  File "/home/grad/pjpark/anaconda3/envs/nequip/lib/python3.8/site-packages/nequip/scripts/train.py", line 81, in main
    trainer = fresh_start(config)
  File "/home/grad/pjpark/anaconda3/envs/nequip/lib/python3.8/site-packages/nequip/scripts/train.py", line 209, in fresh_start
    final_model = model_from_config(
  File "/home/grad/pjpark/anaconda3/envs/nequip/lib/python3.8/site-packages/nequip/model/_build.py", line 133, in model_from_config
    model = builder(**params)
  File "/home/grad/pjpark/anaconda3/envs/nequip/lib/python3.8/site-packages/nequip/model/_scaling.py", line 219, in PerSpeciesRescale
    shifts = torch.as_tensor(shifts)
RuntimeError: Could not infer dtype of dict


If I try

> global_rescale_shift: null
global_rescale_shift_trainable: false
global_rescale_scale: dataset_forces_rms
global_rescale_scale_trainable: false
per_species_rescale_shifts_trainable: false
per_species_rescale_arguments_in_dataset_units: true
per_species_rescale_scales_trainable: true
per_species_rescale_shifts: [Ge:-0.02859140,Sb:-0.03305661,Te:-0.03357790]
per_species_rescale_scales: dataset_per_species_forces_rms


I got 

> Replace string dataset_per_species_forces_rms to tensor([0.4694, 0.6948, 0.6315], dtype=torch.float64)
Traceback (most recent call last):
  File "/home/grad/pjpark/anaconda3/envs/nequip/bin/nequip-train", line 8, in <module>
    sys.exit(main())
  File "/home/grad/pjpark/anaconda3/envs/nequip/lib/python3.8/site-packages/nequip/scripts/train.py", line 81, in main
    trainer = fresh_start(config)
  File "/home/grad/pjpark/anaconda3/envs/nequip/lib/python3.8/site-packages/nequip/scripts/train.py", line 209, in fresh_start
    final_model = model_from_config(
  File "/home/grad/pjpark/anaconda3/envs/nequip/lib/python3.8/site-packages/nequip/model/_build.py", line 133, in model_from_config
    model = builder(**params)
  File "/home/grad/pjpark/anaconda3/envs/nequip/lib/python3.8/site-packages/nequip/model/_scaling.py", line 219, in PerSpeciesRescale
    shifts = torch.as_tensor(shifts)
ValueError: too many dimensions 'str'


It got error when finding 
`Replace string dataset_per_energy_rms to tensor([somthine], dypte-torch.float64)`
the number what I use `[Ge: -0.02859140,Sb: -0.03305661,Te: -0.03357790]` is eV unit

I'm using 0.5.3 v

Do you have any clue?

Thanks

**Reply by pobo95** [reply]
@Linux-cpp-lisp 

Oh It works when I use
`per_species_rescale_shifts: [-0.02859140,-0.03305661,-0.03357790]`

Thank you for your help :)

**Reply by Linux-cpp-lisp** [reply]
Yes:  the shifts should be in order of the assigned atom types, which depend on how you specified the chemical species.

**Reply by pobo95** [reply]
@Linux-cpp-lisp I totally understand thanks :)
